### DISCUSSION TITLE
Trivy not detecting misconfigurations in docker-compose.yml

### DISCUSSION THREAD

**Discussion by matej-kaska** [aquasecurity/trivy#6179] [Q&A] [answered]
### Question

Hello,

I've encountered an issue where Trivy doesn't seem to detect any misconfigurations in my docker-compose.yml file (as it should like in this [example](https://github.com/aquasecurity/trivy/tree/42f795fa34c1d0e0700a941c4771987fd6229b55/examples/misconf/custom-policy/docker-compose/)). I've attempted several approaches to ensure the file is scanned properly, but the output consistently indicates that no config files were detected for scanning.

**Background:**
I'm using Trivy to scan my project for vulnerabilities and misconfigurations, including Dockerfiles and the docker-compose.yml in [GitHub Actions](https://github.com/matej-kaska/ujepsoft/blob/main/.github/workflows/CI.yml). The scans for Dockerfiles work as expected, identifying vulnerabilities and misconfigurations. However, when it comes to the docker-compose.yml file, Trivy doesn't report any findings.

**Steps Taken:**
- Firstly I tried scanning the entire project directory using `trivy fs --security-checks vuln,config .`, expecting it to pick up the docker-compose.yml. The log indicated that the scan was performed, but no config files were detected.
- Secondly I tried scanning just the confg file `trivy config docker-compose.yml`
- Thirdly I created folder "test" and copied the docker-compose.yml to the folder and ran `trivy conf ./test`
- Lastly I tried renaming "docker-compose.yml to "docker-compose.yaml"

**Log Output:**
```
2024-02-21T20:31:56.702Z INFO Misconfiguration scanning is enabled
2024-02-21T20:31:56.702Z INFO Detected config files: 0
```

I confirmed that the docker-compose.yml file exists in the project directory and is correctly formatted. The ls command lists the docker-compose.yml among other project files, ensuring it's in the right place for Trivy to access.

The [docker-compose.yml](https://github.com/matej-kaska/ujepsoft/blob/main/docker-compose.yml) includes various services, volumes, and configurations typical for a development environment, with nothing out of the ordinary in terms of syntax or structure.

Has anyone else experienced similar issues with Trivy not detecting or scanning docker-compose.yml files? Are there specific configurations or considerations I might be missing?

Thank you in advance for your help!

### Target

Filesystem

### Scanner

Misconfiguration

### Output Format

JSON

### Mode

Standalone

### Operating System

ubuntu-latest

### Version

```bash
Version: 0.49.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-02-21 18:10:09.915629643 +0000 UTC
  NextUpdate: 2024-02-22 00:10:09.915629242 +0000 UTC
  DownloadedAt: 2024-02-21 21:16:21.530718269 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-02-21 11:07:56.507504908 +0000 UTC
  NextUpdate: 2024-02-24 11:07:56.507504788 +0000 UTC
  DownloadedAt: 2024-02-21 21:16:54.840259634 +0000 UTC
Policy Bundle:
  Digest: sha256:73a2a1a91c421860d22f08b990a0ca28fee4ca1e1b45e0bdea14357867e31eb6
  DownloadedAt: 2024-02-21 21:17:08.453326829 +0000 UTC
```

**Comment by simar7** [comment, marked as answer]
The example you're looking at is for a custom policy, if defined, will scan for misconfigurations in a docker-compose.yml file. Trivy does not support docker-compose.yml out of the box as documented [here](https://aquasecurity.github.io/trivy/v0.49/docs/coverage/iac/docker/#misconfiguration).

**Comment by HonoSumi** [comment]
Same problem, not resolved. Have tried the custom policy way but no effect. Always shows: INFO    Detected config files   num=0

**Reply by nikpivkin** [reply]
Hi! Can you give an example of your check and how you ran Trivy? You can see examples of custom checks for `docker-compose.yaml` [here](https://github.com/aquasecurity/trivy-checks/tree/main/examples/docker-compose).

**Reply by HonoSumi** [reply]
> Hi! Can you give an example of your check and how you ran Trivy? You can see examples of custom checks for `docker-compose.yaml` [here](https://github.com/aquasecurity/trivy-checks/tree/main/examples/docker-compose).

First I downloaded trivy_0.61.0_Linux-64bit.tar.gz and tar -zxvf it in my /root dir. 
Then I copied the docker-compose.yaml and latest_tag.rego from https://github.com/aquasecurity/trivy-checks/tree/main/examples/docker-compose

Now the dir looks like this:
![image](https://github.com/user-attachments/assets/b821996e-3384-4069-851d-e079cf400418)

After that I typed 
./trivy config --config-check **./latest_tag.rego** --namespaces user **./** -d --skip-check-update
as shown in https://trivy.dev/latest/docs/scanner/misconfiguration/custom/
![image](https://github.com/user-attachments/assets/381b98a1-4900-47b7-aa30-b4e57f5b1fad)

The results turn out to be like:
![image](https://github.com/user-attachments/assets/97569eac-fe4d-432d-b361-d72ececab32c)

The output **_INFO    Detected config files   num=0_** indicates that the docker-compose.yaml is not  scanned.

**Reply by nikpivkin** [reply]
YAML and JSON scanners are disabled by default. See documentation about [YAML scanning](https://trivy.dev/latest/docs/scanner/misconfiguration/#scan-arbitrary-json-and-yaml-configurations).

**Reply by HonoSumi** [reply]
> YAML and JSON scanners are disabled by default. See documentation about [YAML scanning](https://trivy.dev/latest/docs/scanner/misconfiguration/#scan-arbitrary-json-and-yaml-configurations).

Great! Now the scanning works as expected. Many thanks!
BTW, I'm wondering why Trivy does not support the scanning of docker-compose by default, since it is also a frequently used container technology.

**Reply by nikpivkin** [reply]
@simar7 @itaysk Are there any plans to support `docker-compose.yaml`?

**Reply by simar7** [reply]
@nikpivkin I don't see why not - we just never got to it until now. We could enable to be a scan target by default since we already have most of the scan logic anyway.

**Reply by nikpivkin** [reply]
> We could enable to be a scan target by default

You mean enable the scanner for `docker-compose` by default, so that users don't have to do it manually using the flag?

**Reply by simar7** [reply]
> > We could enable to be a scan target by default
> 
> You mean enable the scanner for `docker-compose` by default, so that users don't have to do it manually using the flag?

Yes I think it would be an extension of [this](https://trivy.dev/latest/docs/scanner/misconfiguration/#scan-arbitrary-json-and-yaml-configurations) but somehow we'll have to figure out that it is a docker-compose.yaml and not just a regular yaml file.

Conventionally docker-compose.yaml is named as such but afaik there isn't any requirement to do so.

**Reply by nikpivkin** [reply]
> Yes I think it would be an extension of [this](https://trivy.dev/latest/docs/scanner/misconfiguration/#scan-arbitrary-json-and-yaml-configurations) but somehow we'll have to figure out that it is a docker-compose.yaml and not just a regular yaml file.

we can define docker-compose to match the schema, Trivy supports this

**Reply by simar7** [reply]
Makes sense - track https://github.com/aquasecurity/trivy/issues/8729 cc @HonoSumi
